features/251-templates/index.md

```diff
@@ -2,57 +2,92 @@
-  TemplateConfigurationTypeForgotPasswordEmailHTML            TemplateConfigurationType = "forgot_password_email.html"
-  TemplateConfigurationTypeForgotPasswordResetHTML            TemplateConfigurationType = "forgot_password_reset.html"            // Keep URL option
-  TemplateConfigurationTypeForgotPasswordSuccessHTML          TemplateConfigurationType = "forgot_password_success.html"          // Keep URL option
-  TemplateConfigurationTypeForgotPasswordErrorHTML            TemplateConfigurationType = "forgot_password_error.html"            // Keep URL option
-  TemplateConfigurationTypeWelcomeEmailTXT                    TemplateConfigurationType = "welcome_email.txt"
-  TemplateConfigurationTypeWelcomeEmailHTML                   TemplateConfigurationType = "welcome_email.html"
-  TemplateConfigurationTypeUserVerificationGeneralErrorHTML   TemplateConfigurationType = "user_verification_general_error.html"  // Keep URL option
-  TemplateConfigurationTypeUserVerificationMessageTXT         TemplateConfigurationType = "user_verification_message.txt"
-  TemplateConfigurationTypeUserVerificationMessageHTML        TemplateConfigurationType = "user_verification_message.html"
-  TemplateConfigurationTypeUserVerificationSuccessHTML        TemplateConfigurationType = "user_verification_success.html"        // Keep URL option
-  TemplateConfigurationTypeUserVerificationErrorHTML          TemplateConfigurationType = "user_verification_error.html"          // Keep URL option
-)
+templates/phone/user_verification_message.txt    # this is an SMS message template
 ```
 
 ### Template Language Tag
 
-Each template may optionally have a language tag. In the future, we may support sending localized message. The language tag is specified in [BCP47](https://tools.ietf.org/html/bcp47).
+Each template may optionally have a language tag. The language tag is specified in [BCP47](https://tools.ietf.org/html/bcp47).
 
-### Template Key
+## Template resolution
+
+To resolve a template, the input is the template type, optionally the template key and finally the user preferred languages. The type and key is determined by the feature while the user preferred languages is provided by the user.
+
+All templates have default value so template resolution always succeed.
+
+The templates are first resolved by matching the type and the key. And then select the best language according to the user preferred languages.
+
+## Component templates
+
+Some template may depend on other templates which are included during rendering. This enables customizing a particular component of a template.
+
+For example, `auth_ui_login.html` depend on `auth_ui_header.html` and `auth_ui_footer.html` to provide the header and footer. If the developer just wants to customize the header, they do not need to provide customized templates for ALL pages. They just need to provide `auth_ui_header.html`.
```

Comment:
How to prevent excessive recursion (e.g. billion laugh attack for XML) if we allow user template to include another template?

Reply:
Internally, the template spec must declare the templates they depend on. Since the template spec is written by us, we can check the AST to ensure user cannot include an unlisted template. We should also ensure template do not have cycle.

Reply:
I understand it correctly, that means each template type have a whitelisted of template types that it could include. That is, `auth_ui_login.html` can import only `auth_ui_header.html` and `auth_ui_footer.html`, and `auth_ui_header.html`/`auth_ui_footer.html` cannot import any template?

Reply:
They can, as long as there is no cycle.

Reply:
It's just that we'll need to inspect the content of all templates to check whether the template reference graph is a DAG.

Reply:
I think we do not need to detect cycle with graph. The dependency is static and is code reviewed. At parsing time, we just need to check if the template has `{{template name}}` where `name` is not listed.

Reply:
But template is customizable? e.g. developer can update `auth_ui_login.html` to include `auth_ui_header.html`, and update `auth_ui_header.html` to include `auth_ui_login.html`? Or do we maintain a whitelist of which template can include what templates?

Reply:
> Or do we maintain a whitelist of which template can include what templates?

Yes. The template spec would have its `Defines` replaced by `Dependencies []TemplateItemType`.

Reply:
If the developer wants to change the structure of `auth_ui_login.html` entirely, like instead of `auth_ui_header.html` and `auth_ui_footer.html`, their custom template expects `auth_ui_sidebar.html` and `auth_ui_nav.html` then dependencies must not be declared in the template spec and we need to use graph algorithm to detect cycles.

I think we can leave this to future.